### PR TITLE
ProgressTracker emits exception thrown by the flow, allowing the ANSI…

### PR DESCRIPTION
**Pull request by shamsasari**
… renderer to correctly stop and print the error (#189)

The `ProgressTracker.changes` observable now stops on the same error that the flow throws. This resolves issue #189 so it no longer continues printing to DONE. It now prints a 🚫 for all the steps that didn't occur due to the exception and a ☠️ at the end with the exception message.

The flow exception should now pass down to the `StateMachineInfo.progressTrackerStepAndUpdates` and `FlowHandle.progress` RPC observables.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/utilities/Emoji.kt [hunk 1]
```diff
@@ -16,6 +16,8 @@ object Emoji {
     const val CODE_GREEN_TICK = "\u2705"
     const val CODE_PAPERCLIP = "\ud83d\udcce"
     const val CODE_COOL_GUY = "\ud83d\ude0e"
+    @JvmStatic val CODE_NO_ENTRY = String(Character.toChars(128683))
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/utilities/Emoji.kt]: Why are these inconsistent with the rest? (annotations, const)

File: core/src/main/kotlin/net/corda/core/utilities/ProgressTracker.kt [hunk 2]
```diff
@@ -149,6 +149,8 @@ class ProgressTracker(vararg steps: Step) {
         _changes.onNext(Change.Structural(this, step))
     }
 
+    fun endWithError(error: Throwable) = _changes.onError(error)
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/utilities/ProgressTracker.kt]: Quick API doc? I know it's pretty obvious anyway.

File: core/src/main/kotlin/net/corda/core/utilities/ProgressTracker.kt [hunk 3]
```diff
@@ -195,6 +197,8 @@ class ProgressTracker(vararg steps: Step) {
      * if a step changed its label or rendering).
      */
     val changes: Observable<Change> get() = _changes
+
+    val hasEnded: Boolean get() = _changes.hasCompleted() || _changes.hasThrowable()
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/utilities/ProgressTracker.kt]: ditto

File: samples/trader-demo/src/integration-test/kotlin/net/corda/traderdemo/TraderDemoTest.kt [hunk 4]
```diff
@@ -4,31 +4,35 @@ import com.google.common.util.concurrent.Futures
 import net.corda.core.getOrThrow
```
Comment (mikehearn) [inline samples/trader-demo/src/integration-test/kotlin/net/corda/traderdemo/TraderDemoTest.kt]: I think this file could have been split out into a separate PR, right? It's fine though.

Comment (shamsasari) [inline samples/trader-demo/src/integration-test/kotlin/net/corda/traderdemo/TraderDemoTest.kt]: I made this change to make debugging easier (it's also faster). I can revert back if you wish.

Comment (mikehearn) [inline samples/trader-demo/src/integration-test/kotlin/net/corda/traderdemo/TraderDemoTest.kt]: No no, it's a good change. Just a reminder to try and split unrelated things into separate commits so that e.g. bisection is easier.

File: core/src/main/kotlin/net/corda/core/utilities/Emoji.kt (hunk 1 above)
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/utilities/Emoji.kt]: I'll change all of them to use codepoints.

Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/utilities/Emoji.kt]: Actually, IntelliJ lets us embed Emoji in source code directly now. So I'm not sure we need escaping at all.

**Comment by shamsasari** [comment]
Comments addressed.

File: core/src/main/kotlin/net/corda/core/utilities/Emoji.kt (hunk 1 above)
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/utilities/Emoji.kt]: Copy and pasting the emoji didn't get Intellij to display it.

Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/utilities/Emoji.kt]: ah - it will be in 2017.1, my bad:

https://blog.jetbrains.com/idea/2016/12/intellij-idea-2017-1-eap-is-open/

I thought it was out already.

Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/utilities/Emoji.kt]: Nice. First thing I'll do when it's released!
